Added vertical reading mode to the book reader

CKolle opened the pull request:
# Added
- Added: Vertical reading mode support added to the web-based book reader. This feature is particularly useful for reading genres like Japanese light novels. The API backend and settings user interface have also been updated to allow users to switch between horizontal and vertical reading modes
# Changed
- Changed: Margin handling has been refined. The units were changed from percentage (%) to viewport width (vw) to accommodate resizing when the reading direction is set to right-to-left (rtl).

majora2007 commented on API/Entities/Enums/ReadingMode.cs:
```diff
@@ -0,0 +1,7 @@
+
+public enum ReadingMode
+{
+    Vertically = 0,
```
Add [Description()] annotation like other enums

majora2007 commented on UI/Web/src/app/book-reader/_components/book-reader/book-reader.component.ts:
```diff
@@ -846,19 +884,30 @@ export class BookReaderComponent implements OnInit, AfterViewInit, OnDestroy {
     } else if (scrollTop !== undefined && scrollTop !== 0) {
       this.scrollService.scrollTo(scrollTop, this.reader.nativeElement);
+    } else if ((this.readingMode === ReadingMode.Vertically) && (this.layoutMode === BookPageLayoutMode.Default)) {
+       console.log("Scrolling to end of page");
```
Remove console.log

majora2007 commented on UI/Web/src/app/book-reader/_components/book-reader/book-reader.component.ts:
```diff
@@ -846,19 +884,30 @@ export class BookReaderComponent implements OnInit, AfterViewInit, OnDestroy {
         this.scrollService.scrollTo(0, this.reader.nativeElement);
-      } else {
+      } else if (this.readingMode === ReadingMode.Vertically) {
+        this.reader.nativeElement.children
```
Remove dead code

majora2007 commented on UI/Web/src/app/book-reader/_components/book-reader/book-reader.component.ts:
```diff
@@ -931,7 +980,12 @@ export class BookReaderComponent implements OnInit, AfterViewInit, OnDestroy {
+        } else {
+          this.scrollService.scrollToX((currentVirtualPage - 2) * pageWidth, this.bookContentElemRef.nativeElement);
+        }
+        //this.scrollService.scrollToX((currentVirtualPage - 2) * pageWidth, this.bookContentElemRef.nativeElement);
```
Remove dead code

majora2007 commented on UI/Web/src/app/book-reader/_components/book-reader/book-reader.component.ts:
```diff
@@ -956,14 +1010,17 @@ export class BookReaderComponent implements OnInit, AfterViewInit, OnDestroy {
         // +0 apparently goes forward 1 virtual page...
-        this.scrollService.scrollToX((currentVirtualPage) * pageWidth, this.bookContentElemRef.nativeElement);
+        if (this.readingMode === ReadingMode.Vertically) {
+          this.scrollService.scrollTo( (currentVirtualPage) * pageWidth, this.bookContentElemRef.nativeElement, "auto");
```
Use single quotes in javascript

majora2007 commented on UI/Web/src/app/book-reader/_components/book-reader/book-reader.component.ts:
```diff
@@ -996,36 +1053,46 @@ export class BookReaderComponent implements OnInit, AfterViewInit, OnDestroy {
+    const [scroll, total] = this.getScrollAndTotal();
+    const pageSize = this.getPageSize();
+    const totalVirtualPages = Math.max(1, Math.round(total / pageSize));
+    const currentVirtualPage = Math.min(Math.max(1, Math.round((scroll + pageSize) / pageSize)), totalVirtualPages);
```
Put my comment back because I always forget how virtual page math works.

majora2007 commented on UI/Web/src/app/book-reader/_components/reader-settings/reader-settings.component.ts:
```diff
@@ -160,9 +172,15 @@ export class ReaderSettingsComponent implements OnInit, OnDestroy {
           this.user.preferences.bookReaderReadingDirection = ReadingDirection.LeftToRight;
         }
+        if (this.user.preferences.bookReaderReadingMode === undefined) {
+          console.log('Reading mode is undefined  - setting to horizontal');
```
Remove dead code

majora2007 reviewed (changes requested):
Preliminary glance over looks pretty good. After v0.7 releases I will pull this down and give a indepth review.

majora2007 commented:
I haven't forgotten about this, I will be on holiday until next week. I have this planned for a code review and extensive testing shakeout.

majora2007 commented:
Here are some issues I faced:
-  Vertical/Horizontal is always showing flipped on ebook menu. When I'm in Horizontal, it's showing Vertical as the label. (Edit: I'm just realizing it's because that's how the text flows. I think this may be confusing and we should use how the page is read vs how the text is laid out)
- I cannot get Horizontal + Scroll to work. It's always clamping at 1 column and I have to use Tap to Paginate to move pages. I was finally able to trigger it by going to 2 column then back to scroll (on vertically), but scroll via mouse isn't possible. 

Once it's functionally working, I will do a code review. Overall, it feels really good though and outside of those small issues, this looks like a huge improvement. I really appreciate the PR.

majora2007 reviewed (changes requested):
See comment above

CKolle commented:
I will look into the the issues

CKolle commented:
I have now addressed the issues you faced and squashed some bugs I found regarding pagination in vertical scroll mode, and images getting a little cropped.

majora2007 commented:
I'm getting build errors.

horizontalScrollbarNeeded isn't defined on book-reader.component.ts

CKolle commented:
Sorry about that I forgot to commit a line, I must have thought it was an unnecessary white-space or something. Anyways did a fresh git clone and tested it, and it seems to be working fine.

majora2007 commented on UI/Web/src/app/book-reader/_components/book-reader/book-reader.component.ts:
```diff
@@ -817,19 +861,34 @@ export class BookReaderComponent implements OnInit, AfterViewInit, OnDestroy {
-        this.renderer.setStyle(img, 'max-height', height);
-      });
+    let maxHeight: number | undefined;
+    let columnGap = 20
```
Make this const

majora2007 commented on UI/Web/src/app/book-reader/_components/book-reader/book-reader.component.ts:
```diff
@@ -996,36 +1071,66 @@ export class BookReaderComponent implements OnInit, AfterViewInit, OnDestroy {
+    if (this.readingSectionElemRef == null) return 0;
+    const height = (parseInt(this.ColumnHeight.replace('px', ''), 10));
+
+    const columnGap = 20;
```
Move this and the other into one at the top. We shouldn't have 2 constants in diferent places.

majora2007 commented on UI/Web/src/app/book-reader/_components/book-reader/book-reader.component.ts:
```diff
@@ -996,36 +1071,66 @@ export class BookReaderComponent implements OnInit, AfterViewInit, OnDestroy {
+    if (!this.bookContentElemRef || !this.readingSectionElemRef) return [1, 1, 0];
 
-    const totalVirtualPages = Math.max(1, Math.round((totalScroll) / pageWidth));
+    const [scroll, total] = this.getScrollAndTotal();
```
I'd prefer to keep the variable names scrollOffset and totalScroll

majora2007 commented on UI/Web/src/app/book-reader/_components/book-reader/book-reader.component.ts:
```diff
@@ -1148,7 +1257,12 @@ export class BookReaderComponent implements OnInit, AfterViewInit, OnDestroy {
+    if(this.layoutMode === BookPageLayoutMode.Default && this.writingStyle === WritingStyle.Vertical ) {
+      const windowWidth = window.innerWidth || document.documentElement.clientWidth;
+      const scrollLeft = element.getBoundingClientRect().left + window.pageXOffset - (windowWidth - element.getBoundingClientRect().width);
+      setTimeout(() => this.scrollService.scrollToX(scrollLeft, this.reader.nativeElement, "smooth"), 10);
```
Please use single quotes in the javascript

majora2007 commented on UI/Web/src/app/book-reader/_components/book-reader/book-reader.component.ts:
```diff
@@ -1210,19 +1324,37 @@ export class BookReaderComponent implements OnInit, AfterViewInit, OnDestroy {
+      });
+    } else if (this.bookContentElemRef !== undefined) {
+      const resumeElement: string | null | undefined = this.getFirstVisibleElementXPath();
+      if (resumeElement !== null && resumeElement !== undefined) {
```
this can just be if (resumeElement)

and you can remove the typing from resumeElement as IDE can infer from getFirstVisibleElementXPath

majora2007 commented on UI/Web/src/app/book-reader/_components/reader-settings/reader-settings.component.html:
```diff
@@ -70,6 +70,15 @@ <h2 class="accordion-header">
                     </button>
                 </div>
+                <div class="controls" style="display: flex; justify-content: space-between; align-items: center; ">
+                    <label id="writingstyle" class="form-label">Writing Style <i class="fa fa-info-circle" aria-hidden="true" placement="top" [ngbTooltip]="writingStyleTooltip" role="button" tabindex="0" aria-describedby="writingStyle-help"></i></label>
```
this should be writing-style. For any ids in html, it should be kebab case.

majora2007 replied:
You also need a span with writingStyle-help as an id so the screenreader announces what's in the tooltip. Look around and you should find an example

majora2007 commented on UI/Web/src/app/book-reader/_components/reader-settings/reader-settings.component.html:
```diff
@@ -70,6 +70,15 @@ <h2 class="accordion-header">
                 </div>
+                <div class="controls" style="display: flex; justify-content: space-between; align-items: center; ">
+                    <label id="writingstyle" class="form-label">Writing Style <i class="fa fa-info-circle" aria-hidden="true" placement="top" [ngbTooltip]="writingStyleTooltip" role="button" tabindex="0" aria-describedby="writingStyle-help"></i></label>
+                    <ng-template #writingStyleTooltip>Changes the direction of the text. Horizontal is left to right, vertical is top to bottom.</ng-template>
```
This seems non-intuitive for first time user, but i can't think of a better description right now.

majora2007 commented on UI/Web/src/app/user-settings/user-preferences/user-preferences.component.html:
```diff
@@ -251,6 +252,15 @@ <h2 class="accordion-header">
 
                                 <div class="row g-0">
+                                    <div class="col-md-6 col-sm-12 pe-2 mb-3">
+                                        <label for="settings-book-reading-style" class="form-label">Writing Style</label>&nbsp; <i class="fa fa-info-circle" aria-hidden="true" placement="right" [ngbTooltip]="bookWritingStyleToolTip" role="button" tabindex="0"></i>
```
Don't use nbsp&;, use mr-2 class (search around you'll see the correct class)

majora2007 replied:
You need the aria-describeby="settings-book-writing-style-help" so the screenreader understands the tooltip message

majora2007 reviewed (commented):
Mostly looking good, a few comments.

majora2007 commented on UI/Web/src/app/book-reader/_components/book-reader/book-reader.component.ts:
```diff
@@ -202,6 +202,7 @@ export class BookReaderComponent implements OnInit, AfterViewInit, OnDestroy {
    * Used for showing/hiding bottom action bar. Calculates if there is enough scroll to show it.
    * Will hide if all content in book is absolute positioned
    */
+  columnGap: number = 20;
```
Move this to a const outside the class since it's not changing. Then you don't need to access on this.

majora2007 reviewed (commented):
Almost there

majora2007 reviewed (approved):
LGTM
